### PR TITLE
feat(workflow): Adding Actor model and User/Team FK to Actor

### PR DESCRIPTION
**Pull request by iProgramStuff**
We're adding an "Actor" model as an abstraction to represent owners and teams. In followup PRs, we'll begin writing Actor rows for new Teams and Users and then backfill existing ones.

This PR adds the Actor model and an indexed FK on User and Team that point to it.

Going forward, we'll be able to use this to easily represent an "actor" in a model via:
`actor = FlexibleForeignKey("sentry.Actor")`

and you'll be able to get the Team/User by selecting it based on actor.type where the id's match. Something like
```
select * from sentry_groupassignee ga left join user u on ga.actor_id = u.actor_id left join team t on ga.actor_id = t.actor_id
```
would now be possible.

This will be the new standard at Sentry (in the past we've been using two fields on new models to represent owners) going forward and used in the new AlertRule and Rule owner features. In the long-long-term we may go back and change existing models to use this system.

### CODE REVIEW

File: src/sentry/models/actor.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+class ActorType(Enum):
+    TEAM = 0
+    USER = 1
+
+
+class Actor(Model):
+    __core__ = True
+
+    organization = FlexibleForeignKey("sentry.Organization")
```
Comment (iProgramStuff) [inline src/sentry/models/actor.py:15]: Can I use this to meaningfully speed up the query?

File: src/sentry/migrations/0170_actor_introduction.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# Generated by Django 1.11.29 on 2021-03-02 21:36
+
+from django.db import migrations, models
+import django.db.models.deletion
+import sentry.db.models.fields.bounded
+import sentry.db.models.fields.foreignkey
+import sentry.models.actor
```
Comment (wedamija) [inline src/sentry/migrations/0170_actor_introduction.py:7]: Did Django automatically add this? Surprised it's importing an external file

Comment (markstory) [reply src/sentry/migrations/0170_actor_introduction.py:7]: Isn't using imports in migrations like this risky as the model could move/be removed in the future causing the migration to fail.

File: src/sentry/migrations/0170_actor_introduction.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                (
+                    "type",
+                    models.PositiveSmallIntegerField(
+                        choices=[
+                            (sentry.models.actor.ActorType(0), "team"),
+                            (sentry.models.actor.ActorType(1), "user"),
```
Comment (markstory) [inline src/sentry/migrations/0170_actor_introduction.py:50]: Duplicating the values into the migration will be more future proof as the `ActorType` object could change down the road or be removed.

File: src/sentry/migrations/0170_actor_introduction.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                null=True, on_delete=django.db.models.deletion.CASCADE, to="sentry.Actor"
+            ),
+        ),
+        migrations.AddField(
+            model_name="user",
+            name="actor",
+            field=sentry.db.models.fields.foreignkey.FlexibleForeignKey(
+                null=True, on_delete=django.db.models.deletion.CASCADE, to="sentry.Actor"
+            ),
+        ),
```
Comment (markstory) [inline src/sentry/migrations/0170_actor_introduction.py:78]: Do we need a check constraint to ensure that only one of user or team is populated and prevent the none and both states?

File: src/sentry/migrations/0170_actor_introduction.py (hunk 2 above)
Comment (iProgramStuff) [reply src/sentry/migrations/0170_actor_introduction.py:7]: Yeah this was automatically added. It's using the enum to populate the choices. I can definitely just write those in manually though.

File: src/sentry/migrations/0170_actor_introduction.py (hunk 3 above)
Comment (iProgramStuff) [reply src/sentry/migrations/0170_actor_introduction.py:50]: Wrote the 0/1 in manually.

File: src/sentry/migrations/0170_actor_introduction.py (hunk 4 above)
Comment (iProgramStuff) [reply src/sentry/migrations/0170_actor_introduction.py:78]: I'm not sure I understand

File: src/sentry/models/team.py [hunk 5]
```diff
@@ -111,6 +111,7 @@ class Team(Model):
         ),
         default=TeamStatus.VISIBLE,
     )
+    actor = FlexibleForeignKey("sentry.Actor", db_index=True, null=True, on_delete=models.PROTECT)
```
Comment (wedamija) [inline src/sentry/models/team.py]: Oh yeah, this should be a unique index

**Comment by wedamija** [comment]
I think you'll need to add these unique indexes concurrently. You probably just need to create the column as non-unique, then add a unique index to it concurrently and mark as a constraint

**Review by wedamija — APPROVED** [review]
Lgtm. Let's:
 - mark this as `is_dangerous`
 - have ops run the  migration
 - merge/deploy
 
 This should be fine to be automated, but I'm a little paranoid around the user/team tables so just want to be cautious.

**Pull request merged**